src/core/common/message.cpp

```diff
@@ -143,30 +154,34 @@ void MessagePool::FreeBuffers(Buffer *aBuffer)
     }
 }
 
+#if OPENTHREAD_MTD || OPENTHREAD_FTD
+otError MessagePool::ReclaimBuffers(Message::Priority aPriority)
+{
+    return Get<MeshForwarder>().EvictMessage(aPriority);
+}
+
 otError MessagePool::ReclaimBuffers(int aNumBuffers, Message::Priority aPriority)
 {
-#if OPENTHREAD_MTD || OPENTHREAD_FTD
     while (aNumBuffers > GetFreeBufferCount())
     {
-        SuccessOrExit(Get<MeshForwarder>().EvictMessage(aPriority));
+        SuccessOrExit(ReclaimBuffers(aPriority));
     }
 
 exit:
-#else
-    OT_UNUSED_VARIABLE(aPriority);
-#endif
-
     // First comparison is to get around issues with comparing
     // signed and unsigned numbers, if aNumBuffers is negative then
     // the second comparison wont be attempted.
     return (aNumBuffers < 0 || aNumBuffers <= GetFreeBufferCount()) ? OT_ERROR_NONE : OT_ERROR_NO_BUFS;
 }
+#endif
 
 uint16_t MessagePool::GetFreeBufferCount(void) const
 {
     uint16_t rval;
 
-#if OPENTHREAD_CONFIG_PLATFORM_MESSAGE_MANAGEMENT
+#if OPENTHREAD_CONFIG_MESSAGE_USE_HEAP_ENABLE
+    rval = static_cast<uint16_t>(GetInstance().GetHeap().GetFreeSize() / sizeof(Buffer));
```

Comment:
This is not necessarily correct when using Heap because there might be fragmentation? Even when the free size is larger than sizeof(Buffer), we might still unable to allocate one. 

This also makes the method `MessagePool::ReclaimBuffers(int aNumBuffers, Message::Priority aPriority)` unable to guarantee that there are enough message buffers available.

Reply:
Yes, that's one of the reason removing the use of in SetLength().

Reply:
Yes, that's one of the reasons removing the use of in SetLength().